### DISCUSSION TITLE
Use `Deno.realPathSync()` in an Edge Function to perform dynamic imports.

### DISCUSSION THREAD

**Discussion by anandkkpr** [supabase/supabase#29010] [Questions] [answered]
Hi everyone!

I am working on setting up a factory Class Auto-loader that I want to use to automate some of the drudge work related to Deno.serve() within Edge Functions.

I was thinking of setting up a Class that takes the `Deno.serve(REQUEST)` Request object and then auto-loads the appropriate controller using a JS/TS "dynamice import" to attempt to load a module/class using `await import `${req.method}.controller.ts` to load the module if it exists.

Looks like this effort is stumped because it looks like `Deno.cwd()` only returns `.` when called from `Deno.serve(){}` within the `index.ts` file. When called from another file that is loaded by static import statement (e.g. `index.ts:1` - `import {ClassLoader} from './helpers/classloader.ts'`), `Deno.cwd()` returns nothing (an empty string).

I understand that Deno permissions are involved here. For instance, when I try to run `Deno.realpathSync()`, I get a "PermissionDenied" error.

```shell
[Error] PermissionDenied: Deno.realPathSync is blocklisted
    at Object.realPathSync (ext:sb_core_main_js/js/bootstrap.js:387:9)
    at Object.handler (file:///PATH/TO/EDGE/REPO/supabase/functions/some_edge_function/index.ts:6:33)
    at respond (ext:sb_core_main_js/js/http.js:163:38)
    at handleHttp (ext:sb_core_main_js/js/http.js:123:5)
    at eventLoopTick (ext:core/01_core.js:168:7) {
  name: "PermissionDenied"
}
```

The question here then is:

1. Is there a way to tell Supabase to allow certain Deno permissions to support dynamic path imports in Edge Functions?
2. If not the above, then does anyone have a suggestion for dynamically importing classes in a way that Supabase allows?

Thank you for any assistance/advice you can provide!

**Comment by gc-ft** [comment, marked as answer]
You might not like my answer, but:

1. this is not possible due to limitations in how edge functions are executed and to protect the environment (my guess)
2. suggestion: do not do it :)

Why do I say the above? Edge functions are very short lived and very resource limited. Therefor, each edge function should do AS LITTLE as possible AS FAST as possible AS EFFICIENT as possible.

Creating edge functions with many files, lots of dynamic imports, factory class auto-loaders and other slow inefficient things for something like an edge function (which gets booted multiple times a minute in multiple instances on higher load) is not the way to go in my opinion.

Make small functions, fast to execute, small, packaged well together. This will help boot time and resource limits tremendously.

Fabian

**Reply by anandkkpr** [reply]
Hi @gc-ft, yes, you're the bearer of bad news... But never fear. I thank you for it... I'm glad I have a definitive answer!

I had suspected as much (on the security side of it) but wanted to be sure that I wasn't missing something.

I've re-architected the stuff I'm doing and am using a simple switch statement now based on `request.method`. It's not magical and mistake proof... but overall there is a pattern in my edge function now that can be followed by the team which should allow for better organisation and easier code reviews among other desirable SOLID-like stuff.

Thank you for your time!

**Comment by chmoder** [comment]
`@google-cloud/*` libs don't work because realPathSync is disabled.  For example, [google cloud storage](https://www.npmjs.com/package/@google-cloud/storage) has 3 Million+ downloads per week.  This caused some delays for us and these are popular libraries so I thought you may want to know.  We would appreciate a solution where we can use these libraries with supabase edge functions.
